caas/kubernetes/provider/rbacmapper.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	core "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type RBACMapper interface {
+	worker.Worker
+	AppNameForServiceAccount(types.UID) (string, error)
+}
+
+type rbacMapper struct {
+	catacomb     catacomb.Catacomb
+	lock         *sync.RWMutex
+	logger       loggo.Logger
+	saInformer   core.ServiceAccountInformer
+	saNameUIDMap map[string]types.UID
+	saUIDAppMap  map[types.UID]string
+	workQueue    workqueue.RateLimitingInterface
+}
+
+const (
+	RBACMapperResyncPeriod = time.Minute * 5
+)
+
+func (r *rbacMapper) AppNameForServiceAccount(id types.UID) (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	appName, found := r.saUIDAppMap[id]
+	if !found {
+		return "", errors.NotFoundf("no service account for app found with id %v", id)
+	}
+	return appName, nil
+}
+
+func (r *rbacMapper) enqueueServiceAccount(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		r.logger.Errorf("failed enqueuing service account: %v", err)
+		return
+	}
+	r.workQueue.Add(key)
+}
+
+func (r *rbacMapper) Kill() {
+	r.catacomb.Kill(nil)
+}
+
+func newRBACMapper(logger loggo.Logger, informer core.ServiceAccountInformer) (*rbacMapper, error) {
+	rm := &rbacMapper{
+		lock:         new(sync.RWMutex),
+		logger:       logger,
+		saInformer:   informer,
+		saNameUIDMap: map[string]types.UID{},
+		saUIDAppMap:  map[types.UID]string{},
+		workQueue:    workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &rm.catacomb,
+		Work: rm.loop,
+	}); err != nil {
+		return rm, errors.Trace(err)
+	}
+	return rm, nil
+}
+
+func (r *rbacMapper) loop() error {
+	defer r.workQueue.ShutDown()
+	r.saInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    r.enqueueServiceAccount,
+		DeleteFunc: r.enqueueServiceAccount,
+		UpdateFunc: func(_, newObj interface{}) {
+			r.enqueueServiceAccount(newObj)
+		},
+	})
+
+	go r.saInformer.Informer().Run(r.catacomb.Dying())
+	go wait.Until(func() {
+		for r.processNextQueueItem() {
+		}
+	}, time.Second, r.catacomb.Dying())
```

Comment:
This seems a little wrong. Maybe I'm missing something.
processNextQueueItem() will exit with false when there's an error
this will exit the wait Until() func
then we block on catacomb.Dying() and the worker is stuck
The normal behaviour is to exit the loop() on error and bounce the worker

Reply:
Added a comment to explain.